core/trino-main/src/main/java/io/trino/operator/join/unspilled/HashBuilderOperator.java

```diff
@@ -306,6 +306,11 @@ private void finishInput()
             return;
         }
 
+        ListenableFuture<Void> reserved = localUserMemoryContext.setBytes(index.getEstimatedLookupSourceSizeInBytes(hashArraySizeSupplier));
+        if (!reserved.isDone()) {
+            // Yield when not enough memory is available to proceed, finish is expected to be called again when some memory is freed
```

Comment:
That seems to be awfully dependent on `Driver` detail. Could we make operator return blocked future instead.

Please add tests also

Reply:
> That seems to be awfully dependent on Driver detail. Could we make operator return blocked future instead.

I tend to disagree. It is generally true that currently it is responsibility of the system to "park" Drivers when the system is low on memory. For example if the `HashAggregationOperator` pushes the memory pool over the limit we currently don't explicitly block the operator, instead we simply update the memory utilization and expect the engine not to proceed if the system is low on memory. 

> Please add tests also

Yeah, still working on it.